coding-articles/yehia-article-thegraph/Article.md

```diff
@@ -0,0 +1,660 @@
+
+Create a directory and initialize the project
+
+```
+# Inside the project directory
+graph init
+```
+
+This command will ask you some questions
+
+1. Protocol? Ethereum
+2. Product for which to initialize? Subgraph-studio hosted service may be better if you want to control who queries your graph and the cost
+3. Subgraph slug? I will name it swappairs
+4. Ethereum Network? mainnet
+5. Contract Address? 0x5c69bee701ef814a2b6a3edd4b1652cb9cc5aa6f
+6. Contract Name? Pairs
+   Then the project will start installing, and you should see something like this 👇
+
+<figure>
+  <img src="./images/graph_init.png" alt="Initialize graph output">
+</figure>
+
+After installing, you should now access the project from your code editor.
+
+<figure>
+  <img src="./images/project_root.png" alt="Initialize graph output">
+</figure>
+
+Start with the entities inside schema.graphql and put the data you need to
+First schema is the `Token` it will be simple we need only the contract address and the token name
+
+```
+type Token @entity {
+  id: ID! # contract address
+  name: String!
+}
+```
+
+Second schema is the `CreatedPairs` its the two pair of tokens details, the transaction hash and the timestamp.
+
+```
+type CreatedPair @entity {
+  id: ID! # transaction hash
+  token0: Token!
+  token1: Token!
+  timestamp: BigInt!
+}
+```
+
+Schema file is ready you can check it on [src/schema.graphql](code/subgraph/schema.graphql) last thing is to generate the schema entities by running
+
+```
+yarn codegen
+```
+
+command I will explain it in details later.
+
+Lets do the logic now that
+Go to [src/swap-pairs.ts](code/subgraph/schema.graphql) delete everything inside the file
+
+Import the schema file
+
+```
+// Import the created swap entity
+import { PairCreated } from "../generated/SwapPairs/SwapPairs";
+```
+
+Create the first handler
+
+```
+// PairCreated is an event inside the smart contract
+export function handlePairCreated(event: PairCreated): void {
+    // What will happen after the event is triggered
+}
+```
+
+Inside the `handlePairCreated` function lets initialize the entity for `CreatedPair`
+
+```
+  // Inside handlePairCreated
+  let createdPair = new CreatedPair(event.transaction.hash.toHexString());
+```
+
+the entity take the schema ID as parameter `new CreatedPair(<transaction hash>)`.
+
+Assign values to the schema properties
+
+```
+createdPair.timestamp = event.block.timestamp;
+```
+
+We need to get the two pairs of tokens' names. Unfortunately, the event includes the address of the tokens only, not their names. So we need to read from the smart contract the token name.
+To do this first, let's import ERC20 smart contract ABIs
+
+go to [abis/](./code/subgraph//abis/) and add [Erc20.json](./code/subgraph/abis/Erc20.json) abi file, then write on your terminal/cmd
+
+```
+# inside the project directory
+yarn codegen
+```
+
+That's all now. All initialize the ERC20 smart contract object
+
+```
+  // In src/swap-pairs.ts inside hanldePaireCreated
```

Comment:
Ok two things.

1: First we need to tell people to import relevant objects before we invoke them.
2: the codegen command is not generating the typescript file associated with the ABI automatically for me. This breaks the code for me I guess. 
What could I be doing different?

Reply:
Ending this review iteration here. Looking forward to hearing from you. This is a relatively complicated blog, I expected errors but you did a great job with very few bugs. 
Thanks a lot for your contribution.
I'll help you as best I can to complete this blog perfectly.

Reply:
Thank you @Genesis3800  for your support 🙏 
1. I updated them
2. Try using `yarn`; ensure you are on the root folder of the graph. If not let's have a discord call to check the error and the none rendering images

Reply:
Thank you for your improvements. Lemme go through everything one more time. Will go with yarn this time. 
I'll try to avoid it, but we'll probably need a coding session to patch up everything.

Its' your fault though, for making the article so good :)

Reply:
hahahahaha Thank you @Genesis3800 Looking forward to talk to you 😄

Reply:
Hey @yehia67 ,
it looks like we need to talk over a meet.

1: yarn codegen doesn't work for me. I think its a local yarn error on my machine. 
2: npm run codegen works, and graph codegen does the same thing. 

Even after I run the commands in step 2, my ts file isn't generated from the ABI file. 
Please lemme know if you are free sometime in the coming days, let us solve this properly over a coding session.